[FMHA] Batch Prefill Support Improvements:  Change KV Cache Layout & Large Page Size Support

Jeff-Huang opened the pull request:
## Proposed changes
This PR introduces significant improvements to the Batch Prefill kernels in ck_tile, enabling codegen support, change the KV layouts and support 1024 page size.

 Codegen Support: Enabled code generation for batch_prefill kernels.
       * Updated fmha_batch_prefill.py to generate kernels with different page sizes 1024.
       * Pipeline & Problem Structure:
       * Introduced BlockFmhaBatchPrefillPipelineProblem structure for better management of  batch prefill problems.
       * Updated block_fmha_batch_prefill_pipeline_qr_ks_vs_async.hpp to handle async operations and offsets
         correctly.
   * KV Cache & Layout Enhancements:
       * Added support for k cache layout: [num_blocks, num_kv_heads, head_size/8, block_size, 8].
       * Added support for v cache layout: [num_blocks, num_kv_heads, block_size/8, head_size, 8].
       * Added kv_last_page_lens to kernel arguments to support handling varying sequence lengths in paged attention.
       * Refined kv_offset_array_transform to calculate k/v offset.
       * Added support for large page size (1024).
   * Bug Fixes:
       * Fixed incorrect page ID calculation in kv_offset_array_transform specifically for gfx950.
       
## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [ ] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

ltqin committed:
add page_block_size parameter

ltqin committed:
add is_sglang_layout to  parameters

ltqin committed:
add kv_offset_array_transform to batch async for page size 16

ltqin committed:
add kv_last_page_lens to kernel

ltqin committed:
change kv layout to [num_total_pages, page_block_size, hdim]

ltqin committed:
format

Jeff-Huang committed:
- enable codegen of batch_prefill kernels
- create new problem struct BlockFmhaBatchPrefillPipelineProblem for
  batch prefill kernels
- generate different page sizes of batch prefill kernels (1, 16)

Jeff-Huang committed:
1. fix wrong calculation of page id in kv_offset_array_transform in gfx950
2. support page size 1024

Jeff-Huang committed:
fix python format

Jeff-Huang committed:
change kv cache layout to [num_blocks, num_kv_heads, head_size/x,
block_size, x] and [num_blocks, num_kv_heads, block_size/X, head_size, X]

poyenc commented on example/ck_tile/01_fmha/codegen/ops/fmha_batch_prefill.py:
```diff
@@ -526,17 +545,19 @@ def template(self) -> str:
         # TODO: we don't encode idx here
         return (
-            f"fmha_batch_prefill_d{self.F_hdim}_{self.F_dtype}_{self.F_mode}_"
+            f"fmha_batch_prefill_d{self.F_hdim}_{self.F_dtype}_{self.F_mode}_p{self.F_page_size}_"
```
suggest to use "ps" here

poyenc commented on example/ck_tile/01_fmha/fmha_fwd.hpp:
```diff
@@ -1272,6 +1266,60 @@ struct fmha_fwd_traits_
+          bool kSkipMinSeqlenQ_            = false,
+          bool kIsSglangLayout_            = false,
+          ck_tile::index_t kPageBlockSize_ = 1>
+struct fmha_fwd_batch_prefill_traits_
```
suggest to derive from fmha_fwd_traits_

poyenc commented on example/ck_tile/01_fmha/fmha_fwd.hpp:
```diff
@@ -1518,7 +1566,24 @@ float fmha_fwd_appendkv(fmha_fwd_appendkv_traits,
                         const ck_tile::stream_config&);
 
-using fmha_batch_prefill_traits = fmha_fwd_traits;
+struct fmha_batch_prefill_traits
```
suggest to derive from fmha_fwd_traits

poyenc commented on include/ck_tile/ops/fmha/kernel/fmha_batch_prefill_kernel.hpp:
```diff
@@ -101,6 +103,7 @@ struct FmhaBatchPrefillWithPagedKVCacheKernel
             "v" + (std::is_same_v<VLayout, ck_tile::tensor_layout::gemm::RowMajor> ? "r" : "c") + (pn.empty() ? "_npad" : "_" + pn) +
             (kHasLogitsSoftCap ? "_logits" : "_nlogits" ) + (BiasEnum == BlockAttentionBiasEnum::NO_BIAS ? _SS_("_nbias") : (_SS_("_") + BlockAttentionBiasEnumToStr<BiasEnum>::name)) +
             (kHasMask ? "_" + _SS_(FmhaMask::name) : "_nmask") + (kStoreLSE ? "_lse" : "_nlse" ) + (kHasDropout ? "_dropout" : "_ndropout" ) +
+            (kIsSglangLayout ? "_sglang" : "_vllm" ) + (_SS_("_pg") + _TS_(kPageBlockSize)) + 
```
typo? use "ps" here

poyenc commented on include/ck_tile/ops/fmha/kernel/fmha_batch_prefill_kernel.hpp:
```diff
@@ -821,58 +812,96 @@ struct FmhaBatchPrefillWithPagedKVCacheKernel
                 k_ptr,
-                make_tuple(kargs.num_total_pages * kargs.page_block_size, kargs.hdim_q),
-                make_tuple(kargs.stride_k, 1),
+                make_tuple(kargs.num_total_pages, kargs.hdim_q / 8, kargs.page_block_size, 8),
```
use `16 / sizeof(KDataType)` here. avoid the hardcoded 8

poyenc commented on include/ck_tile/ops/fmha/kernel/fmha_batch_prefill_kernel.hpp:
```diff
@@ -821,58 +812,96 @@ struct FmhaBatchPrefillWithPagedKVCacheKernel
+            //    Strides: (BatchStride, HeadDim*8, 8, 1)
+            const auto v_dram_naive = make_naive_tensor_view<address_space_enum::global>(
+                v_ptr,
+                make_tuple(kargs.num_total_pages, kargs.page_block_size / 8, kargs.hdim_v, 8),
```
use `16 / sizeof(VDataType)` here

poyenc commented on include/ck_tile/ops/fmha/pipeline/block_fmha_batch_prefill_pipeline_qr_ks_vs_async.hpp:
```diff
@@ -12,6 +12,77 @@
+          index_t kLoopStride,
+          bool kIsSglangLayout,
+          bool kIsKcache,
+          index_t kVectorSize = 8>
```
too-specific default value, suggest to not using it and let user decide it

poyenc commented on include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp:
```diff
@@ -65,6 +65,74 @@ struct BlockFmhaPipelineProblem
+          bool kUseTrLoad_,
+          int kPageBlockSize_,
+          typename Traits_>
+struct BlockFmhaBatchPrefillPipelineProblem
```
suggest to derive from BlockFmhaPipelineProblem to avoid duplicated code

poyenc commented on include/ck_tile/ops/fmha/pipeline/tile_fmha_traits.hpp:
```diff
@@ -40,6 +41,38 @@ struct TileFmhaTraits
+          bool kSkipMinSeqlenQ_   = false, /* skip min seqlen q while chunked prefill */
+          bool kIsSglangLayout_   = false,
+          index_t kPageBlockSize_ = 1>
+struct TileFmhaBatchPrefillTraits
```
suggest to derive from TileFmhaTraits to avoid duplicated codes

Jeff-Huang committed:
1. Introduced `kVectorSize` in BlockFmhaBatchPrefillPipelineProblem instead of using hardcode values
2. Makes batch prefill kernel traits structures inherent from fmha fwd
   traits
3. Add some static check for Page size, vector size, hdim, ..., etc.

poyenc committed:
Merge branch 'develop' into ck_tile/batch_prefill_paged_size_16_rebase

Jeff-Huang committed:
[Refactor] Replace is_sglang_layout with Enums for KV cache configuration

Refactored `fmha_batch_prefill` to use `BlockAttentionKVCacheMemoryLayoutEnum` (VECTORIZED/LINEAR) and `BlockAttentionKVCacheLookupTableEnum` (SGLANG_1D/VLLM_2D) instead of a single
boolean.

**Changes:**
*   Added Enum definitions in `block_attention_kvcache_layout_enum.hpp`.
*   Updated Kernel, Pipeline, and Traits to template on these Enums.
*   Implemented `kv_offset_array_transform` logic based on `kKVMemoryLayout`.
*   Refactored `PageBlockTableKargs` to adapt to `kKVLookupTable`.
*   Updated CodeGen scripts to support new parameters.

This decouples memory layout from the paging mechanism, enabling flexible KV cache configurations.

Jeff-Huang committed:
Merge remote-tracking branch 'origin/develop' into ck_tile/batch_prefill_paged_size_16_rebase

Jeff-Huang committed:
fix vgpr spill count

poyenc commented on include/ck_tile/ops/fmha/kernel/fmha_batch_prefill_kernel.hpp:
```diff
@@ -56,12 +57,23 @@ struct FmhaBatchPrefillWithPagedKVCacheKernel
 
     static constexpr bool kUseAsyncCopy = FmhaPipeline::Policy::AsyncCopy;
 
+    // clang-format off
```
we no-longer need those templates.

poyenc commented on include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp:
```diff
@@ -65,6 +66,85 @@ struct BlockFmhaPipelineProblem
+                                      kUseTrLoad_,
+                                      Traits_>
+{
+    using QDataType             = remove_cvref_t<QDataType_>;
```
Do we really need those type ailas? they are already declared in `BlockFmhaPipelineProblem<>`

Jeff-Huang replied:
The duplicated alias are removed in last commit.

poyenc commented on example/ck_tile/01_fmha/codegen/ops/fmha_batch_prefill.py:
```diff
@@ -230,12 +248,15 @@ class FmhaFwdApiTrait:
         return (
             f"{self.hdim}-{self.dtype}-{self.mode}-{self.bm0}-{self.bn0}-{self.bk0}-{self.bn0}-{self.bk1}-{self.bk0max}-"
-            + f"{self.vlayout}-{self.logits}-{self.mask}-{self.bias}-{self.lse}-{self.dropout}-{self.qscale}-{self.spad}-{self.skpad}-{self.dpad}-{self.dvpad}"
+            + f"{self.vlayout}-{self.logits}-{self.mask}-{self.bias}-{self.lse}-{self.dropout}-{self.qscale}-{self.spad}-{self.skpad}-{self.dpad}-{self.dvpad}-{self.kv_memory_layout}-{self.kv_lookup_table}-p{self.page_size}"
```
Use prefix "ps" for consistency

poyenc commented on example/ck_tile/01_fmha/codegen/ops/fmha_batch_prefill.py:
```diff
@@ -604,23 +647,43 @@ def get_pipelines(dtype, hdim, receipt, mask_impl) -> List[FmhaFwdPipeline]:
+                SUPPORTED_KV_LOOKUP_TABLE,
             ):
-                pipelines.append(FmhaFwdPipeline("qr_async", "row", "t", "t", "t", "t", logits, bias, lse, dropout, qscale, mask))  # fmt: skip
+                pipelines.append(FmhaFwdPipeline("qr_async", "row", "t", "f", "t", "t", logits, bias, lse, dropout, qscale, mask, kv_memory_layout, kv_lookup_table))  # fmt: skip
```
why do we need this pipeline? there is no point to use "f" for KV seqlen padding

Jeff-Huang replied:
The pipeline with sk_pad=False is removed in last commit.

poyenc commented on include/ck_tile/ops/fmha/kernel/fmha_batch_prefill_kernel.hpp:
```diff
@@ -809,58 +847,111 @@ struct FmhaBatchPrefillWithPagedKVCacheKernel
             }
         }();
         const auto k_dram = [&]() {
+            // New K Layout: [NumPages, D/kVectorSize, S, kVectorSize]
```
I guess this tensor view is not compatible with the old layout?

Jeff-Huang replied:
Yes, this kernel only supports the VECTORIZED_LAYOUT temporarily.

Jeff-Huang committed:
1. remove batch prefill pipeline with sk_pad=false
2. correct some comments
3. add static assert to make sure v offsets is in same page within a tile.

Jeff-Huang committed:
Merge branch 'develop' into ck_tile/batch_prefill_paged_size_16_rebase

Jeff-Huang commented on include/ck_tile/ops/fmha/pipeline/block_fmha_batch_prefill_pipeline_qr_ks_vs_async.hpp:
```diff
@@ -6,12 +6,82 @@
+            {
+                kv_offset_vec[k0] = page_loc + static_cast<long_index_t>(page_offset) * stride_kv;
+            }
+        });
```
We assume no page-boundary crossings; which implies that kPageBlockSize must be divisible by kN0. have also added a static assertion in block_fmha_batch_prefill_pipeline_qr_ks_vs_async.hpp  to ensure this assumption is not violated.

`static_assert(kPageBlockSize % kN0 ==  #0,
                  "V offset assumes each tile stays within a page; kPageBlockSize must be "
                  "divisible by kN0.");`

Jeff-Huang committed:
remove unnecessary t2s functions

Jeff-Huang committed:
add fp8 support for receipt 200 and 600 in fmha_bath_prefill.py

Jeff-Huang committed:
support linear kv cache layout

Jeff-Huang committed:
Remove block_table_ptr from fwd_batch_prefill_args. Instead, reuse
kv_page_indices as a pointer of the lookup table.

Jeff-Huang committed:
Merge branch 'develop' into ck_tile/batch_prefill_paged_size_16_rebase

poyenc commented on include/ck_tile/ops/fmha/kernel/fmha_batch_prefill_kernel.hpp:
```diff
@@ -809,60 +839,177 @@ struct FmhaBatchPrefillWithPagedKVCacheKernel
+                    make_tuple(sequence<2>{}, sequence<0>{}, sequence<1>{}),
+                    make_tuple(sequence<0>{}, sequence<1>{}, sequence<2>{}));
+
+                auto v_dram_final = transform_tensor_view(
```
could we call the `transform_tensor_view()` only once?

Jeff-Huang replied:
sure! I merge these transforms into a single transform in last commit.

Jeff-Huang committed:
1. merge multiple transforms into single transform.
2. add static check to make sure vlayout is row-major.

Jeff-Huang committed:
Merge branch 'develop' into ck_tile/batch_prefill_paged_size_16_rebase

poyenc commented on include/ck_tile/ops/fmha/kernel/fmha_batch_prefill_kernel.hpp:
```diff
@@ -89,14 +112,9 @@ struct FmhaBatchPrefillWithPagedKVCacheKernel
-        static constexpr ck_tile::index_t page_block_size = 1;
-#endif
+        PageBlockTableKargs page_table;
+        const int32_t* seqlen_k_ptr;
```
Is the `seqlen_k_ptr` acctually part of `VllmPageTableKargs`?

Jeff-Huang replied:
yes, seqlen_k_ptr belong to VllmPageTableKargs. I moved it into VllmPageTableKargs in last commit.

Jeff-Huang committed:
move FmhaFwdCommonKargs::seqlen_k_ptr to VllmPageTableKargs.

afagaj commented:
@Jeff-Huang This is a significant PR.  

Please add one or more entries to CHANGELOG.md as appropriate.

Jeff-Huang committed:
update changelog

Jeff-Huang committed:
Merge branch 'develop' into ck_tile/batch_prefill_paged_size_16_rebase

Jeff-Huang committed:
Merge branch 'develop' into ck_tile/batch_prefill_paged_size_16_rebase

Jeff-Huang commented:
> @Jeff-Huang This is a significant PR.
> 
> Please add one or more entries to CHANGELOG.md as appropriate.

Thanks for the reminder! I've updated the CHANGELOG.md with the relevant entries in the latest commit.
